lib/active_model/serializer/fragment_cache.rb

```diff
@@ -0,0 +1,99 @@
+module ActiveModel
+  class Serializer
+    class FragmentCache
+
+      attr_reader :serializer
+
+      def initialize(adapter, serializer, options, root)
+        @root       = root
+        @options    = options
+        @adapter    = adapter
+        @serializer = serializer
+      end
+
+      def fetch
+        klass = serializer.class
+        serializers = fragment_serializer(@serializer.object.class.name, klass)
+
+        cached_serializer = serializers[:cached].constantize.new(@serializer.object)
+        non_cached_serializer = serializers[:non_cached].constantize.new(@serializer.object)
+
+        cached_hash = @adapter.class.new(cached_serializer, @options).serializable_hash
+        non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+        if @serializer.root && @adapter.class == ActiveModel::Serializer::Adapter::JsonApi
+          cached_hash_root(cached_hash, non_cached_hash)
+        else
+          non_cached_hash.merge cached_hash
+        end
+      end
+
+      private
+
+      def cached_hash_root(cached_hash, non_cached_hash)
+        hash = {}
+
+        core_cached = cached_hash.first
+        core_non_cached = non_cached_hash.first
+        no_root_cache = cached_hash.delete_if {|key, value| key == core_cached[0] }
+        no_root_non_cache = non_cached_hash.delete_if {|key, value| key == core_non_cached[0] }
+
+        if @root
+          hash[@root] = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
+        else
+          hash = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
+        end
+        hash.merge no_root_non_cache.merge no_root_cache
+      end
+
+      def fragment_associations(serializers, associations)
+        associations.each do |association|
+          options = ",#{association[1][:association_options]}" if association[1].include?(:association_options)
+          eval("#{serializers[:non_cached]}.#{association[1][:type].to_s}(:#{association[0]}#{options})")
```

Comment:
Commenting. No there is no need, I just forgot it there while I was reviewing the code, I should have replaced it with ```.send```